rest/blip_client_test.go

```diff
@@ -65,19 +71,229 @@ type BlipTesterClient struct {
 	nonCollectionAwareClient *BlipTesterCollectionClient
 }
 
+func (c *BlipTesterCollectionClient) getClientDocForSeq(seq clientSeq) (*clientDoc, bool) {
+	c.seqLock.RLock()
+	defer c.seqLock.RUnlock()
+	doc, ok := c._seqStore[seq]
+	return doc, ok
+}
+
+// OneShotDocsSince is an iterator that yields client sequence and document pairs that are newer than the given since value.
+func (c *BlipTesterCollectionClient) OneShotDocsSince(ctx context.Context, since clientSeq) iter.Seq2[clientSeq, *clientDoc] {
+	return func(yield func(clientSeq, *clientDoc) bool) {
+		c.seqLock.Lock()
+		seqLast := c._seqLast
+		for c._seqLast <= since {
+			if ctx.Err() != nil {
+				c.seqLock.Unlock()
+				return
+			}
+			// block until new seq
+			c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - waiting for new sequence", since, c._seqLast)
+			c._seqCond.Wait()
+			// Check to see if we were woken because of Close()
+			if ctx.Err() != nil {
+				c.seqLock.Unlock()
+				return
+			}
+			seqLast = c._seqLast
+			c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - woke up", since, c._seqLast)
+		}
+		c.seqLock.Unlock()
+		c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - iterating", since, seqLast)
+		for seq := since; seq <= seqLast; seq++ {
+			doc, ok := c.getClientDocForSeq(seq)
+			// filter non-latest entries in cases where we haven't pruned _seqStore
+			if !ok {
+				continue
+			} else if latestDocSeq := doc.latestSeq(); latestDocSeq != seq {
+				// this entry should've been cleaned up from _seqStore
+				base.AssertfCtx(base.TestCtx(c.TB()), "seq %d found in _seqStore but latestSeq for doc %d - this should've been pruned out!", seq, latestDocSeq)
+				continue
+			}
+			if !yield(seq, doc) {
+				c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - stopping iteration", since, seqLast)
+				return
+			}
+		}
+		c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - done", since, seqLast)
+	}
+}
+
+func (c *BlipTesterCollectionClient) docsSince(ctx context.Context, since clientSeq, continuous bool) chan *clientDoc {
+	ch := make(chan *clientDoc)
+	c.goroutineWg.Add(1)
+	go func() {
+		defer c.goroutineWg.Done()
+		sinceVal := since
+		for {
+			if ctx.Err() != nil {
+				close(ch)
+				return
+			}
+			c.TB().Logf("docsSince: sinceVal=%d", sinceVal)
+			for _, doc := range c.OneShotDocsSince(ctx, sinceVal) {
+				select {
+				case <-ctx.Done():
+					close(ch)
+					return
+				case ch <- doc:
+					c.TB().Logf("sent doc %q to changes feed", doc.id)
+					sinceVal = doc.latestSeq()
+				}
+			}
+			if !continuous {
+				c.TB().Logf("opts.Continuous=false, breaking changes loop")
+				break
+			}
```

Comment:
readability nit here:

```suggestion
			defer close(ch)
			for _, doc := range c.OneShotDocsSince(ctx, sinceVal) {
				select {
				case <-ctx.Done():
					return
				case ch <- doc:
					sinceVal = doc.latestSeq()
				}
			}
			if !continuous {
				break
			}
```

Without this change, I don't understand how the loop closes in the non continuous case.

Reply:
It closes from the context when that is cancelled, but otherwise doesn't.

The loop will sit waiting in the `OneShotDocsSince` iterator until a new sequence is available on the client via the `c._seqCond.Wait()`